File: app/models/wish.py
"""Wish ORM model."""

import uuid
from datetime import datetime

from geoalchemy2 import Geography
from sqlalchemy import CheckConstraint, DateTime, ForeignKey, Index, Integer, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class Wish(Base):
    __tablename__ = "wishes"
    __table_args__ = (
        CheckConstraint("status IN ('active','closed','expired')", name="wishes_status_check"),
        Index("ix_wishes_user_id", "user_id"),
        Index("ix_wishes_category_id", "category_id"),
        Index("ix_wishes_status", "status"),
        Index("ix_wishes_expires_at", "expires_at"),
        Index("ix_wishes_location", "location", postgresql_using="gist"),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, server_default=func.gen_random_uuid()
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False
    )
    category_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("categories.id", ondelete="CASCADE"), nullable=False
    )
    title: Mapped[str] = mapped_column(Text, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    location = mapped_column(Geography("POINT", srid=4326), nullable=False)
    location_name: Mapped[str | None] = mapped_column(Text, nullable=True)
    city: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(Text, nullable=False, server_default="active")
    expires_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    max_participants: Mapped[int] = mapped_column(Integer, nullable=False, server_default="1")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

File: frontend/src/types/wish.ts
import type { GeoLocation } from "./geo";

export type WishStatus = "active" | "closed" | "expired";

export interface Category {
  id: string;
  name: string;
  slug: string;
}

export interface Wish {
  id: string;
  user_id: string;
  category_id: string;
  title: string;
  description: string | null;
  location: GeoLocation;
  status: WishStatus;
  max_participants: number;
  spots_left: number | null;
  expires_at: string | null;
  created_at: string;
}

export interface WishCreatePayload {
  title: string;
  description: string | null;
  category_id: string;
  location: GeoLocation;
  max_participants: number;
  expires_at: string | null;
}

export type WishUpdatePayload = Partial<WishCreatePayload> & {
  status?: "active" | "closed";
};

export interface WishFeedQuery {
  near: { latitude: number; longitude: number; radius_km: number };
  category_id?: string;
  limit?: number;
  offset?: number;
}
